File: 0622-design-circular-queue/0622-design-circular-queue.py
class MyCircularQueue:

    def __init__(self, k: int):
        self.array = [0] * k
        self.front = 0
        self.back = 0
        self.size = 0 

    def enQueue(self, value: int) -> bool:
        if self.size == len(self.array):
            return False
        
        self.array[self.back] = value
        self.size += 1
        self.back = (self.back + 1) % (len(self.array))
        return True

    def deQueue(self) -> bool:
        if self.size == 0:
            return False
        
        self.front = (self.front + 1) % (len(self.array))
        self.size -= 1
        return True

    def Front(self) -> int:
        if self.size == 0:
            return -1

        return self.array[self.front]

    def Rear(self) -> int:
        if self.size == 0:
            return -1
        
        return self.array[(self.back - 1) % len(self.array)]

    def isEmpty(self) -> bool:
        return self.size == 0

    def isFull(self) -> bool:
        return self.size == len(self.array)


# Your MyCircularQueue object will be instantiated and called as such:
# obj = MyCircularQueue(k)
# param_1 = obj.enQueue(value)
# param_2 = obj.deQueue()
# param_3 = obj.Front()
# param_4 = obj.Rear()
# param_5 = obj.isEmpty()
# param_6 = obj.isFull()

File: 0622-design-circular-queue/0622-design-circular-queue.swift

class MyCircularQueue {
    let capacity: Int
    var arr: [Int]
    var front = 0
    var rear = 0
    var size = 0

    init(_ k: Int) {
        capacity = k
        arr = Array(repeating: 0, count: k)
    }
    
    func enQueue(_ value: Int) -> Bool {
        if size == capacity {
            return false
        }

        arr[rear] = value
        rear = (rear + 1) % capacity
        size += 1

        return true
    }
    
    func deQueue() -> Bool {
        if size == 0 {
            return false
        }

        front = (front + 1) % capacity
        size -= 1

        return true
    }
    
    func Front() -> Int {
        if size == 0 {
            return -1
        }

        return arr[front]
    }
    
    func Rear() -> Int {
        if size == 0 {
            return -1
        }

        if rear == 0 {
            return arr.last!
        }

        return arr[rear - 1]
    }
    
    func isEmpty() -> Bool {
        return size == 0
    }
    
    func isFull() -> Bool {
        return size == capacity
    }
}

/**
 * Your MyCircularQueue object will be instantiated and called as such:
 * let obj = MyCircularQueue(k)
 * let ret_1: Bool = obj.enQueue(value)
 * let ret_2: Bool = obj.deQueue()
 * let ret_3: Int = obj.Front()
 * let ret_4: Int = obj.Rear()
 * let ret_5: Bool = obj.isEmpty()
 * let ret_6: Bool = obj.isFull()
 */